workflow-manager/limiter/limiter.go

```diff
@@ -0,0 +1,69 @@
+package limiter
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// Limiter is similar to sync.Pool, however, it limtis how many shared
+// objects (in this case workers) there can be
+type Limiter struct {
+	workers    chan int32
+	inProgress int32
+}
+
+// Ticket is the structure that allows you to schedule jobs
+type Ticket struct {
+	limiter *Limiter
+	worker  int32
+	dirty   int32
+}
+
+// New creates a new instance of Limiter that limits how many functions can get executed without the other ones completing first
+func New(maxWorkers int32) *Limiter {
+	l := &Limiter{make(chan int32, maxWorkers), 0}
+
+	for i := int32(0); i < maxWorkers; i++ {
+		l.workers <- i
+	}
+
+	return l
+}
+
+// RequestTicket requests a ticket for a worker
+// this call will block if there are no tickets available for work
+func (l *Limiter) RequestTicket() *Ticket {
+	worker := <-l.workers
+	return &Ticket{
+		limiter: l,
+		worker:  worker,
+		dirty:   int32(0),
+	}
+}
+
+// Execute Executes a function with the work ticket
+func (t *Ticket) Execute(job func()) error {
+	if atomic.LoadInt32(&t.dirty) > 0 {
+		return fmt.Errorf("did not schedule job - ticket was dirty")
+	}
+
+	atomic.AddInt32(&t.dirty, 1)
+	atomic.AddInt32(&t.limiter.inProgress, 1)
+	job()
+
+	return nil
+}
+
+// CompleteWork signals that the execution for this particular task is complete
+func (t *Ticket) Done() error {
```

Comment:
nit/subjective: something feels odd to me about `Ticket` having a `Limiter` member since you get a `Ticket` from a `Limiter`, so it feels like parent/child relationship is being inverted here. I wonder if having a method `(l *Limiter) Done(ticket *Ticket) error` would work better?

Reply:
Hmm, yeah. I kinda feel like that makes more sense but at the same time it feels like the "Ticket" is done. Maybe it would make more sense to pass the worker back to the limiter on some sort of channel.